lib/services/metadata_provider.dart

```diff
@@ -0,0 +1,140 @@
+import 'package:finamp/models/finamp_models.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:get_it/get_it.dart';
+import 'package:logging/logging.dart';
+
+import '../models/jellyfin_models.dart';
+import 'downloads_service.dart';
+import 'finamp_settings_helper.dart';
+import 'jellyfin_api_helper.dart';
+
+final metadataProviderLogger = Logger("MetadataProvider");
+
+class MetadataRequest {
+  const MetadataRequest({
+    required this.item,
+    this.includeLyrics = false,
+  }) : super();
+
+  final BaseItemDto item;
+
+  final bool includeLyrics;
+
+  @override
+  bool operator ==(Object other) {
+    return other is MetadataRequest &&
+        other.includeLyrics == includeLyrics &&
+        other.item.id == item.id;
+  }
+
+  @override
+  int get hashCode => Object.hash(item.id, includeLyrics);
+}
+
+enum MetadataState {
+  loading,
+  loaded,
+}
+
+class MetadataProvider {
+
+  MetadataState state = MetadataState.loading;
+  final BaseItemDto item;
+  LyricDto? lyrics;
+
+  MetadataProvider({
+    required this.item,
+    this.lyrics,
+  });
+
+  bool get hasLyrics => item.mediaStreams?.any((e) => e.type == "Lyric") ?? false;
+
+}
+
+final AutoDisposeFutureProviderFamily<MetadataProvider?, MetadataRequest>
+    metadataProvider = FutureProvider.autoDispose.family<MetadataProvider?, MetadataRequest>((ref, request) async {
+
+  final jellyfinApiHelper = GetIt.instance<JellyfinApiHelper>();
+  final downloadsService = GetIt.instance<DownloadsService>();
+
+  metadataProviderLogger.fine("Fetching metadata for '${request.item.name}' (${request.item.id})");
+
+  BaseItemDto? itemInfo;
+  
+  //!!! only use offline metadata if the app is in offline mode
+  // Finamp should always use the server metadata when online, if possible
+  if (FinampSettingsHelper.finampSettings.isOffline) {
+    try {
+      final downloadItem = await downloadsService.getSongInfo(id: request.item.id);
+      if (downloadItem != null) {
+        metadataProviderLogger.fine("Got offline metadata for '${request.item.name}'");
+        itemInfo = downloadItem.baseItem!;
+      }
+    } catch (e) {
```

Comment:
getSongInfo doesn't really have any reasonable failure modes.  You probably don't need a catch block here.

Reply:
It would fail if the item id is null, but I guess catching or not makes little difference in that case...